core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -345,16 +345,15 @@ final case class NonEmptyList[+A](head: A, tail: List[A]) extends NonEmptyCollec
   override def distinctBy[B](f: A => B)(implicit O: Order[B]): NonEmptyList[A] = {
     implicit val ord: Ordering[B] = O.toOrdering
 
-    val buf = ListBuffer.empty[A]
-    tail.foldLeft(TreeSet(f(head): B)) { (elementsSoFar, a) =>
-      val b = f(a)
-      if (elementsSoFar(b)) elementsSoFar
-      else {
-        buf += a; elementsSoFar + b
-      }
+    val bldr = List.newBuilder[A]
+    val seen = mutable.TreeSet.empty[B]
+    val it = iterator
+    while (it.hasNext) {
+      val next = it.next()
+      if (seen.add(f(next)))
+        bldr += next
     }
-
-    NonEmptyList(head, buf.toList)
+    NonEmptyList.fromListUnsafe(bldr.result())
```

Comment:
I think it would be slightly better to do:

```scala
var rest = tail
seen.add(f(head))
while (rest.nonEmpty) {
  val next = rest.head
  rest = rest.tail
  if (seen.add(f(next)) {
    bldr += next
  }
}

NonEmptyList(head, bldr.result())
```

the main changes:

1. we don't have to build an iterator for the list, we just directly iterate on List
2. we don't have to allocate the outermost List just to destructure it inside NonEmptyList.fromListUnsafe.

Reply:
I'd also suggest to check if the source NeList has 1 item only (`tail.isEmpty` or use `sizeCompare`) and simply return `this` in that case.
It would help to avoid allocating unnecessary structures: the builder and tree set and recreating 1-item NeList.

This small optimization could also be applied to the other two collections.